compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUPackSharedMemoryAlloc.cpp

```diff
@@ -0,0 +1,209 @@
+                                 group.allocs.end());
+        mergeGroup.liveness.insert(group.liveness.begin(),
+                                   group.liveness.end());
+        // For simplicity we leave the group empty and don't remove it.
+        group.allocs.clear();
+        group.liveness.clear();
+      }
+    }
+  }
+
+  LLVM_DEBUG({
+    for (size_t i = 0; i < groups.size(); i++) {
+      llvm::dbgs() << "Alias group " << i << ":\n";
+      for (Operation *op : groups[i].allocs) op->dump();
+    }
+  });
+
+  for (size_t i = 0; i < groups.size(); i++) {
+    if (groups[i].allocs.empty()) continue;
+    aliasGroups.push_back(std::move(groups[i].allocs));
+  }
+}
+
+/// Insert barriers and wait operations if there are allocs of a different alias
+/// group before the given alloc.
+static void addBarrier(func::FuncOp funcOp, Operation *alloc) {
+  Block *entryBlock = &(*funcOp.getBlocks().begin());
+  bool needBarrier = false;
+  if (alloc->getBlock() != entryBlock) {
+    needBarrier = true;
+  } else {
+    for (Operation &op : entryBlock->getOperations()) {
+      op.dump();
+      if (&op == alloc) break;
+      if (op.getNumRegions() != 0) {
+        needBarrier = true;
+        break;
+      }
+      if (isa<memref::AllocOp>(&op)) {
+        if (!op.hasAttr(kAllocAliasGroup) ||
+            op.getAttr(kAllocAliasGroup) != alloc->getAttr(kAllocAliasGroup)) {
+          needBarrier = true;
+          break;
+        }
+      }
+    }
+  }
+  if (!needBarrier) return;
+  OpBuilder builder(alloc);
+  // TODO: make it a option if needed.
+  bool hasAsyncCopies = true;
+  if (hasAsyncCopies) {
+    Value groupToken = builder.create<nvgpu::DeviceAsyncCreateGroupOp>(
+        funcOp.getLoc(), nvgpu::DeviceAsyncTokenType::get(funcOp.getContext()),
+        SmallVector<Value>());
+    builder.create<nvgpu::DeviceAsyncWaitOp>(funcOp.getLoc(), groupToken,
+                                             builder.getI32IntegerAttr(0));
+  }
+  builder.create<gpu::BarrierOp>(alloc->getLoc());
+}
+
+namespace {
+
+struct LLVMGPUPackSharedMemoryAllocPass
+    : public LLVMGPUPackSharedMemoryAllocBase<
+          LLVMGPUPackSharedMemoryAllocPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<nvgpu::NVGPUDialect>();
+  }
+
+  void runOnOperation() override {
+    func::FuncOp funcOp = getOperation();
+    DominanceInfo dominators(funcOp);
+    SmallVector<Operation *> allocs;
+    funcOp.walk([&](memref::AllocOp alloc) { allocs.push_back(alloc); });
+    // First sink the alloc as low as possible in the CFG.
+    sinkAlloc(allocs, dominators);
+    SmallVector<SmallVector<Operation *>> aliasGroups;
+    analyzeSharedMemoryAlloc(funcOp, allocs, aliasGroups);
+    // If there is 1 or less alias group there is nothing to do.
+    if (aliasGroups.size() <= 1) return;
+
+    // Idealy we should pack the different allocations into one large alloc that
```

Comment:
@benvanik this is the comment explaining why we cannot do direct transformation.